aten/src/ATen/native/cuda/MemoryAccess.cuh

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <cstdint>
+#include <type_traits>
+#include <c10/util/Exception.h>
+
+// References:
+// https://devblogs.nvidia.com/cuda-pro-tip-increase-performance-with-vectorized-memory-access/
+
+namespace at { namespace native { namespace memory {
+
+// aligned vector generates vectorized load/store on CUDA
+template<typename scalar_t, int vec_size>
+struct alignas(sizeof(scalar_t) * vec_size) aligned_vector {
+  scalar_t val[vec_size];
+};
+
+template <
+  int num_threads_,        // number of threads in a block.
+  int thread_work_size_    // number of elements each block needs to handle.
+>
+struct policies {
+
+  struct common {
+    static constexpr int num_threads = num_threads_;
+    static constexpr int thread_work_size = thread_work_size_;
+    static constexpr int block_work_size = thread_work_size_ * num_threads_;
+  };
+
+  struct checked_unroll : public common {
+
+    int remaining;
+
+    __device__ checked_unroll(int remaining): remaining(remaining) {}
+
+    template<typename accessor_t, typename scalar_t>
+    __device__ inline void load(accessor_t to, scalar_t *from) {
+      int thread_idx = threadIdx.x;
+      #pragma unroll
+      for (int i = 0; i < thread_work_size_; i++) {
+        if (thread_idx >= remaining) {
+          return;
+        }
+        to(i) = from[thread_idx];
+        thread_idx += num_threads_;
+      }
+    }
+
+    template<typename accessor_t, typename scalar_t>
+    __device__ inline void store(scalar_t *to, accessor_t from) {
```

Comment:
Accessor being the first argument to load and the second argument to store seems a little odd?

Reply:
fixed